### PR TITLE
Dataimporter functionality completed and tested, Task-refactoring completed

### PR DESCRIPTION
**Pull request by Decitizen**
* DB initialization by json with new `DataImporter`-class
* Pre-made data available in json format
* Task and TaskImplementation-classes refactored to support individual task-types
* DB-restructured, support for Cascade ALL removed (for now)
* JavaClassGenerator functionality improved and tests added
   * support for filename-generation added

### CODE REVIEW

File: src/main/java/pingis/entities/Challenge.java [hunk 1]
```diff
@@ -36,33 +36,33 @@
     @Max(LEVEL_MAX_VALUE)
     private int level;
     private float rating;
+    private ChallengeType type;
 
     @NotEmpty
-    @OneToMany(fetch=FetchType.LAZY, cascade = {CascadeType.ALL}, mappedBy="challenge")
+    @OneToMany(fetch=FetchType.EAGER, /*cascade = {CascadeType.ALL},*/ mappedBy="challenge")
```
Comment (dwarfcrank) [inline src/main/java/pingis/entities/Challenge.java]: Remove the parameter instead of commenting it out.

File: src/main/java/pingis/entities/Challenge.java [hunk 2]
```diff
@@ -36,33 +36,33 @@
...
     private List<Task> tasks;
 
-    @OneToMany(fetch=FetchType.LAZY, cascade = {CascadeType.ALL}, mappedBy = "challenge")
+    @OneToMany(fetch=FetchType.LAZY, /*cascade = {CascadeType.ALL},*/ mappedBy = "challenge")
```
Comment (dwarfcrank) [inline src/main/java/pingis/entities/Challenge.java]: Same here.

File: src/main/java/pingis/entities/ChallengeImplementation.java [hunk 3]
```diff
@@ -35,7 +35,7 @@
     @ManyToOne(fetch=FetchType.EAGER)
     private User implementationUser;
     
-    @OneToMany(fetch=FetchType.LAZY, cascade = {CascadeType.ALL}, mappedBy = "challengeImplementation")
+    @OneToMany(fetch=FetchType.LAZY, /* cascade = {CascadeType.ALL} ,*/ mappedBy = "challengeImplementation")
```
Comment (dwarfcrank) [inline src/main/java/pingis/entities/ChallengeImplementation.java]: And here.

File: src/main/java/pingis/entities/Task.java [hunk 4]
```diff
@@ -51,17 +54,19 @@
     @ManyToOne(fetch=FetchType.EAGER)
     private User author;
 
-    @OneToMany(fetch=FetchType.LAZY, cascade = {CascadeType.ALL}, mappedBy="task")
+    @OneToMany(fetch=FetchType.LAZY, /*cascade = {CascadeType.ALL},*/ mappedBy="task")
```
Comment (dwarfcrank) [inline src/main/java/pingis/entities/Task.java]: And here.

File: src/main/java/pingis/services/DataImporter.java [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+import pingis.repositories.ChallengeRepository;
+import pingis.repositories.TaskImplementationRepository;
+import pingis.repositories.TaskRepository;
+import pingis.repositories.UserRepository;
+
+@Component
+public class DataImporter implements ApplicationRunner{
+
+    private String jsonString;
+    private ChallengeRepository cr;
```
Comment (dwarfcrank) [inline src/main/java/pingis/services/DataImporter.java]: Don't use acronyms as field names, use `challengeRepository` instead of `cr`.

File: src/main/java/pingis/services/DataImporter.java [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+    public ArrayList<TaskImplementation> getTaskImplementations() {
+        return taskImplementations;
+    }
+
+    public ArrayList<ChallengeImplementation> getChallengeImplementations() {
+        return challengeImplementations;
+    }
+    
+    public void generateEntities() {
+        JSONObject obj = new JSONObject(this.jsonString);
```
Comment (dwarfcrank) [inline src/main/java/pingis/services/DataImporter.java]: Why not use a JSON data binding library like Jackson here instead of parsing it manually?

File: src/main/java/pingis/utils/JavaClassGenerator.java [hunk 7]
```diff
@@ -23,6 +29,20 @@ private static String generateTestClassHeader(Challenge c) {
         return "public class " + c.getName().replaceAll("\\s+","") + "Test {";
     }
     
+    public static String generateImplClassName(Challenge challenge) {
```
Comment (dwarfcrank) [inline src/main/java/pingis/utils/JavaClassGenerator.java]: This method generates a file name rather than a class name.

File: src/main/java/pingis/utils/JavaClassGenerator.java [hunk 8]
```diff
@@ -23,6 +29,20 @@ private static String generateTestClassHeader(Challenge c) {
...
+        String classFilename = returnCorrectLength(challenge.getName());
+        return "src/" + classFilename.replaceAll("[\\d\\s+]","") + ".java";
+    }
+    
+    public static String generateTestClassName(Challenge challenge) {
+        String classFilename = returnCorrectLength(challenge.getName());
+        return "test/" + classFilename.replaceAll("[\\d\\s+]","") + "Test.java";
+    }
+    
+    private static String returnCorrectLength(String string) {
```
Comment (dwarfcrank) [inline src/main/java/pingis/utils/JavaClassGenerator.java]: Why is this needed? Why truncate the file name?

File: src/main/java/pingis/utils/JavaClassGenerator.java [hunk 9]
```diff
@@ -31,7 +51,7 @@ private static String indentateTaskAsMethod(List<Task> tasks) {
         String codeSegment = "";
         
         for (Task task : tasks) {
-            String[] lines = task.getCode().split(System.getProperty("line.separator"));
+            String[] lines = task.getCodeStub().split(System.getProperty("line.separator"));
```
Comment (dwarfcrank) [inline src/main/java/pingis/utils/JavaClassGenerator.java:57]: `codeStub` seems to always be an array in JSON, but contains either everything on a single line or each line as a separate string. Why?

File: src/main/resources/templates/user.html [hunk 10]
```diff
@@ -25,56 +25,41 @@
...
-                    <h5>History</h5>
-                    ...<br/>
-                    ...<br/>
+                History:<br/>
+                <div class="row challengeitem">
+                     Some old challenge<br/>
                 </div>
 
             </div>
-            <div class="col-md-8">
```
Comment (dwarfcrank) [inline src/main/resources/templates/user.html:49]: Why are these changes included in the PR?

File: src/main/java/pingis/entities/ChallengeType.java [hunk 11]
```diff
@@ -0,0 +1,6 @@
+
+package pingis.entities;
+
+public enum ChallengeType {
```
Comment (dwarfcrank) [inline src/main/java/pingis/entities/ChallengeType.java:4]: Why is this included in the PR?

**Comment by dwarfcrank** [comment]
In the future, try to keep pull requests small. Don't do major rewrites or add different types of functionality in one PR.

File: src/main/resources/templates/user.html (hunk 10 above)
Comment (lramirez) [reply src/main/resources/templates/user.html:49]: no idea :o
